[addons] change binary addon system to have prepared for hidden functions / some other cleanups

AlwinEsch opened the pull request:
## Description

With this, the API orgy is drawing to a close, only for few types after the hidden is default need changes and 1-2 request until end.

The main reason for this is to fix the export defined in the macro "ADDONCREATOR", which only referred to Windows.
This means that addons created in the future will be suitable and all other functions in cmake with `set (CMAKE_CXX_VISIBILITY_PRESET hidden)` and `set (CMAKE_VISIBILITY_INLINES_HIDDEN YES)` can be avoided.
This avoids called conflicts if the same thing is due to two different addons or in Kodi.

In addition, a few other little things are used.

Commit 1:
---------------------------
**[addons][vfs] fix "C" API by add of missing "struct" on value**

There was on one function value the for "C" needed "struct" before related structure name missing. For C++ it is OK without only mandatory for use on "C".

Commit 2:
---------------------------
**[addons] add on 3 "C" related headers the ifdef's for header include**

On three headers was still the ifdef's about alternative of "pragma once" missing. This add them.

Is good to have as "pragma once" not a standardized part and possible that not supported everywhere, the "ifdef" instead works always.

This only be on "C" related headers as them maybe base for other languages.

Commit 3:
---------------------------
**[addons] add new dll export definitions**

- ATTRIBUTE_DLL_IMPORT (to import external lib functions)
- ATTRIBUTE_DLL_EXPORT (to export lib functions and have available outside)
- ATTRIBUTE_DLL_LOCAL (to set functions as private and not available outside of Dll)

This done to prevent problems about wrongly exported parts and conflicts that another addon takes wrong places. Related fix to have whole addon (except needed parts) private, where comes on another request.

Before was inside the addon main macro "ADDONCREATOR" only the windows export used. As there now becomes ATTRIBUTE_DLL_EXPORT used is it exported on all OS.

Commit 4:
---------------------------
**[addons] remove unused GetStatus call**

Them was since very long time not needed and mainly on begin by PVR addons to check for e.g. connection or something else.

This remove it, only to have no min version increase is inside addon_base function structure it still inside and need remove in future.

Commit 5:
---------------------------
**[addons] remove all no more needed DllExport functions**

This removes some functions where not needed anymore.

Only stays in:
- ADDON_Create
- ADDON_GetTypeVersion
- ADDON_GetTypeMinVersion

The rest is done via the structure with function addresses.

Also is now after a long time a documentation added to the ADDONCREATOR macro.

Commit 6:
---------------------------
**[addons] improve some documentations**

Before was some parts missing and not good shown inside Doxygen, specially related to CAddonBase and functions with AddonBase.h.

Thie improve it and to have more usable.

There still parts where can be better and from here one step for it.

Commit 7:
---------------------------
**[addons] increase global API version to 1.3.0**

The Min stays the same and older addons still usable.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?

Still need to test on addons!
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][vfs] fix "C" API by add of missing "struct" on value

There was on one function value the for "C" needed "struct" before related
structure name missing. For C+* it is OK without only mandatory for use on "C".

AlwinEsch committed:
[addons] add on 3 "C" related headers the ifdef's for header include

On three headers was still the ifdef's about alternative of "pragma once"
missing. This add them.

Is good to have as "pragma once" not a standardized part and possible that
not supported everywhere, the "ifdef" instead works always.

This only be on "C" related headers as them maybe base for other languages.

AlwinEsch committed:
[addons] add new dll export definitions

This add:
- ATTRIBUTE_DLL_IMPORT (to import external lib functions)
- ATTRIBUTE_DLL_EXPORT (to export lib functions and have available outside)
- ATTRIBUTE_DLL_LOCAL (to set functions as private and not available outside of Dll)

This done to prevent problems about wrongly exported parts and conflicts that
another addon takes wrong places. Related fix to have whole addon (except needed parts)
private, where comes on another request.

Before was inside the addon main macro "ADDONCREATOR" only the windows export used.
As there now becomes ATTRIBUTE_DLL_EXPORT used is it exported on all OS.

AlwinEsch committed:
[addons] remove unused GetStatus call

Them was since very long time not needed and mainly on begin by PVR
addons to check for e.g. connection or something else.

This remove it, only to have no min version increase is inside addon_base
function structure it still inside and need remove in future.

AlwinEsch committed:
[addons] remove all no more needed DllExport functions

This removes some functions where not needed anymore.

Only stays in:
- ADDON_Create
- ADDON_GetTypeVersion
- ADDON_GetTypeMinVersion

The rest is done via the structure with function addresses.

Also is now after a long time a documentation added to the ADDONCREATOR macro.

AlwinEsch committed:
[addons] improve some documentations

Before was some parts missing and not good shown inside Doxygen,
specially related to CAddonBase and functions with AddonBase.h.

Thie improve it and to have more usable.

There still parts where can be better and from here one step for it.

AlwinEsch committed:
[addons] increase global API version to 1.3.0

The Min stays teh same and older addons still usable.

garbear commented on xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/screensaver.h:
```diff
@@ -8,6 +8,9 @@
 
 #pragma once
```
Why not drop `#pragma once` now that we have include guards?

AlwinEsch replied:
Have removed the `#pragma once` as them really no more needed, thanks for the note about.

phunkyfish reviewed (approved):
LGTM

AlwinEsch committed:
[addons] remove "#pragma once" from "C" headers

As there are "#ifdef" included are the "#pragma once" not needed.

The "once" is not a standardized value and can still be possible that
a language can not use it. For this cases the "#ifdef" was added.

AlwinEsch commented:
I bring this in shortly before inputstream request come in.

The bit bad thing here is that Kodi with this can still work with older addons. But if a older Kodi takes a newer addon becomes it critical as then there parts missing.

phunkyfish commented:
> I bring this in shortly before inputstream request come in.
> 
> The bit bad thing here is that Kodi with this can still work with older addons. But if a older Kodi takes a newer addon becomes it critical as then there parts missing.

Luckily we are still in Alpha!

phunkyfish commented:
jenkins build this please
